control-plane/pkg/reconciler/channel/v2/channelv2.go

```diff
@@ -0,0 +1,260 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v2
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corelisters "k8s.io/client-go/listers/core/v1"
+
+	v1 "knative.dev/eventing/pkg/apis/duck/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/kmeta"
+	"knative.dev/pkg/reconciler"
+	"knative.dev/pkg/system"
+
+	messagingv1beta1 "knative.dev/eventing-kafka/pkg/apis/messaging/v1beta1"
+
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/config"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/contract"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/receiver"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/base"
+	"knative.dev/eventing-kafka-broker/control-plane/pkg/reconciler/kafka"
+
+	internals "knative.dev/eventing-kafka-broker/control-plane/pkg/apis/internals/kafka/eventing"
+	internalscg "knative.dev/eventing-kafka-broker/control-plane/pkg/apis/internals/kafka/eventing/v1alpha1"
+	internalsclient "knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/clientset/versioned"
+	internalslst "knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/listers/eventing/v1alpha1"
+	coreconfig "knative.dev/eventing-kafka-broker/control-plane/pkg/core/config"
+	kafkalogging "knative.dev/eventing-kafka-broker/control-plane/pkg/logging"
+)
+
+const (
+	// TopicPrefix is the Kafka Channel topic prefix - (topic name: knative-messaging-kafka.<channel-namespace>.<channel-name>).
+	TopicPrefix          = "knative-messaging-kafka"
+	DefaultDeliveryOrder = internals.Ordered
+
+	KafkaChannelConditionConsumerGroup    apis.ConditionType = "ConsumerGroup" //condition is registered by controller
+	KafkaChannelConditionConfigMapUpdated apis.ConditionType = "ConfigMapUpdated"
+)
+
+var (
+	conditionSet = apis.NewLivingConditionSet(
+		KafkaChannelConditionConsumerGroup,
+		KafkaChannelConditionConfigMapUpdated,
+	)
+)
+
+type Reconciler struct {
+	*base.Reconciler
+	*config.Env
+	ConfigMapLister     corelisters.ConfigMapLister
+	ConsumerGroupLister internalslst.ConsumerGroupLister
+	InternalsClient     internalsclient.Interface
+}
+
+func (r *Reconciler) ReconcileKind(ctx context.Context, channel *messagingv1beta1.KafkaChannel) reconciler.Event {
```

Comment:
I think we're missing some bits:

- [Probing and setting the channel address](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L268-L283)
- [CM tracking](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L116)
- [Topic creation (with and without auth)](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L121-L182)
- [Annotation update](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L229-L246)
- [Finalizer](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L292)

Overall we need pretty much the same logic of the other reconciler, the logic that changes is [reconcileSubscribers](https://github.com/knative-sandbox/eventing-kafka-broker/blob/3f3f37ca49a64f6d3181b4296ce8c2b7b39e7028/control-plane/pkg/reconciler/channel/channel.go#L205-L207)

IMO, if we can refactor the old channel and in a way that we can reuse the logic and plug our own `reconcileSubscribers` that would be better.